perf(api): add asyncio yields to simulators

sfoster1 opened the pull request:
# Overview

The hardware simulators all work by doing extremely minimal work in their leaf functions. This is good and fast, except that it means that those leaf functions often don't ever actually yield to the asyncio loop. Tasks only yield when they hit a file io (including some asyncio functions, like sleep(), that defer back to the loop's select call). That means that if you simulate a long enough protocol, the task that runs the protocol won't ever yield back to the loop, and will starve other tasks.

To fix this, drop in a bunch of await asyncio.sleep(0) in the simulator leaf function calls, which should defer back to the loop and fix the issue.

# Test Plan
This is a simulator-only change so we don't have to actually run protocols, which is nice, but we should try to upload some.

I'm concerned a bit about performance here; defering back to the loop may make simulating take longer. We should:

- [x] Simulate all-module protocols on the robot with and without this and check performance
- [x] Simulate a very long protocol and see if it starves the loop


# Changelog

- Add a bunch of `asyncio.sleep(0)` to simulator leaf functions

# Review requests

- is this appropriate?

# Risk assessment

Low except from performaance.

sfoster1 committed:
refactor(api): add asyncio yields to simulators

The hardware simulators all work by doing extremely minimal work in
their leaf functions. This is good and fast, except that it means that
those leaf functions often don't ever actually yield to the asyncio
loop. Tasks only yield when they hit a file io (including some asyncio
functions, like sleep(), that defer back to the loop's select call).
That means that if you simulate a long enough protocol, the task that
runs the protocol won't ever yield back to the loop, and will starve
other tasks.

To fix this, drop in a bunch of await asyncio.sleep(0) in the simulator
leaf function calls, which should defer back to the loop and fix the issue.

SyntaxColoring commented on api/src/opentrons/drivers/heater_shaker/simulator.py:
```diff
@@ -1,8 +1,13 @@
 from opentrons.drivers.types import Temperature, RPM, HeaterShakerLabwareLatchStatus
 
 
+async def _yield() -> None:
+    await asyncio.sleep(0)
```
Should this move up a level to be shared within `opentrons.drivers`? Then it can have a docstring explaining why all the leaf methods within `opentrons.drivers` should use it. I'm worried the underlying motivation will be lost to time.

You could even make it a decorator if you're feeling fancy.

sfoster1 replied:
that's probably a good idea

SyntaxColoring commented on api/src/opentrons/drivers/heater_shaker/simulator.py:
```diff
@@ -24,53 +29,65 @@ async def is_connected(self) -> bool:
 
     async def open_labware_latch(self) -> None:
         self._labware_latch_state = HeaterShakerLabwareLatchStatus.IDLE_OPEN
+        await _yield()
```
Do `connect()`, `disconnect()`, and `is_connected()` also need an `await _yield()`?

sfoster1 replied:
I think it would be a consistency thing that doesn't make a ton of difference because they don't get called in tight loops and at least `disconnect()` and `is_connected()` have no need to be async anyway, but maybe it's good for consistency's sake

SyntaxColoring commented on api/src/opentrons/drivers/mag_deck/simulator.py:
```diff
@@ -37,13 +45,16 @@ async def disconnect(self) -> None:
 
     async def enter_programming_mode(self) -> None:
-        pass
+        await _yield()
```
Do `connect()` and `disconnect()` also need an `await _yield()`?

SyntaxColoring commented on api/src/opentrons/drivers/mag_deck/simulator.py:
```diff
@@ -9,21 +11,27 @@
 
     async def move(self, location: float) -> None:
         self._height = location
+        await _yield()
```
Soft style suggestion:

Maybe we should consistently make `await _yield()` the first line of the function, so it's always in the same place and so it'll pop out more when something accidentally omits it.

sfoster1 replied:
I think, on a similar level of soft preference, that it should be as close to the end of the function as possible because typically these functions do some data processing and then end with the the calls to the things that in non-simulating mode yield to the loop, so having the yield at the end of the function makes it more like the non-simulating kind. This may go a little better if it's in a decorator.

SyntaxColoring commented on api/src/opentrons/drivers/temp_deck/simulator.py:
```diff
@@ -18,12 +23,15 @@ def __init__(self, sim_model: Optional[str] = None):
         self._temp = Temperature(target=None, current=23)
+        await _yield()
 
     async def connect(self) -> None:
         pass
```
Do `connect()`, `is_connected()`, and `disconnect()` need an `await _yield()`?

SyntaxColoring commented on api/src/opentrons/drivers/thermocycler/simulator.py:
```diff
@@ -30,14 +35,18 @@ async def is_connected(self) -> bool:
 
```
Do `connect()`, `disconnect()`, and `is_connected()` need an `await _yield()`?

SyntaxColoring commented on api/src/opentrons/drivers/thermocycler/simulator.py:
```diff
@@ -49,38 +58,46 @@ async def set_plate_temperature(
     async def deactivate_all(self) -> None:
         await self.deactivate_lid()
         await self.deactivate_block()
+        await _yield()
```
Technically, we don't need this one here because `self.deactivate_lid()` and `self.deactivate_block()` already yield internally. I'm down to leave it in for consistency and refactor safety, though.

SyntaxColoring commented on api/src/opentrons/hardware_control/backends/ot3simulator.py:
```diff
@@ -67,6 +67,12 @@
 log = logging.getLogger(__name__)
 
 
+async def _yield() -> None:
+    # if an await chain doesn't eventually reach a call that returns to the loop, the task
+    # won't yield. The 0-time sleep will yield for a spin without sleeping.
+    await asyncio.sleep(0)
```
Same gist as my other comment: should we share this between `ot3simulator` and `simulator`?

SyntaxColoring commented on api/src/opentrons/hardware_control/backends/ot3simulator.py:
```diff
@@ -406,6 +426,7 @@ async def set_active_current(self, axis_currents: OT3AxisMap[float]) -> None:
+        await _yield()
         return None
 
     @asynccontextmanager
```
This is kind of fun—should this async context manager have an `await _yield()`, if we're being consistent?

```python
    @asynccontextmanager
    async def restore_current(self) -> AsyncIterator[None]:
        """Save the current."""
        await _yield()
        yield  # heh
```

sfoster1 replied:
Lol. I think it does not need to precisely _because_ it is an async context manager, so it only exists to wrap other async hardware controller calls that _do_ have yields

SyntaxColoring commented on api/src/opentrons/hardware_control/backends/ot3simulator.py:
```diff
@@ -478,24 +502,29 @@ def resume(self) -> None:
+        await _yield()
         return {}
 
     async def clean_up(self) -> None:
         """Clean up."""
+        await _yield()
         pass
```
```suggestion
    async def clean_up(self) -> None:
        """Clean up."""
        await _yield()
```

SyntaxColoring commented on api/src/opentrons/hardware_control/backends/simulator.py:
```diff
@@ -199,12 +207,14 @@ async def home(self, axes: Optional[List[str]] = None) -> Dict[str, float]:
 
     async def fast_home(self, axis: Sequence[str], margin: float) -> Dict[str, float]:
         for ax in axis:
             self._position[ax] = self._smoothie_driver.homed_position[ax]
```
Commenting on a random line because GitHub:

`drivers.smoothie_drivers.simulator.SimulatingDriver` also has a few non-yielding `async` methods. Should those have `await _yield()` calls, too?

sfoster1 replied:
No, because the simulator.py has the yields. These don't _have to_ have to be in leaf calls.

I'd really also like to keep all the yields in the same place as much as I can. For the modules, that has to be the drivers because that's where the simulation occurs. For the robot controllers, if it was in the smoothie driver then it would have to be somewhere unrelated for OT3, and I don't really like that.

So, that's what I went with.

SyntaxColoring commented on api/src/opentrons/hardware_control/backends/ot3simulator.py:
```diff
@@ -419,6 +440,7 @@ async def watch(self, loop: asyncio.AbstractEventLoop) -> None:
             for idx, mod in enumerate(self._stubbed_attached_modules)
         ]
         await self.module_controls.register_modules(new_mods_at_ports=new_mods_at_ports)
+        await _yield()
```
It looks like the analogous `watch()` method in `simulator.py` is missing this `await _yield()` addition.

SyntaxColoring reviewed (commented):
Thank you! This code mostly looks good to me. I'll get back to you with some test results.

sfoster1 committed:
Update api/src/opentrons/hardware_control/backends/ot3simulator.py

Co-authored-by: Max Marrone <max@opentrons.com>

sfoster1 reviewed (commented):
I think you're probably right about sharing it... it just annoyingly seems like the right place to share it is 1. in the HC backend and 2. in the driver level, I really wish it was just one place.

sfoster1 committed:
refactor: add async yield helpers

These helpers can be used to make sure that leaf calls in an async tight
loop actually end up yielding to the loop so we don't accidentally
starve other running tasks of execution time.

sfoster1 committed:
refactor: use the helpers

sfoster1 committed:
use paramspecs when typechecking

sfoster1 committed:
Revert "use paramspecs when typechecking"

This reverts commit 5e783a5d230ac74df2c5453047411a4c8c1a78fe.

sfoster1 committed:
remove unused imports, fix typing

SyntaxColoring reviewed (approved):
Sweet, thanks.

Still one open question about the Smoothie `SimulatingDriver`: https://github.com/Opentrons/opentrons/pull/12023#discussion_r1080627020



But this looks good to me to merge if you don't think any action is needed on that.

To sort of answer the performance overhead question: I did some testing in https://github.com/Opentrons/opentrons/pull/12038 and found that the addition of `asyncio.sleep(0)` there didn't make much of a difference. I wouldn't expect the additions of `asyncio.sleep(0)` in this PR to be any different.
